{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Aqui é feito a importação dos dados ncbi_saqua, separados os ids de embl e refseq.\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "\n",
    "\n",
    "def create_chunks(list_name, n):\n",
    "    for i in range(0, len(list_name), n):\n",
    "        yield list_name[i:i + n]\n",
    "\n",
    "df_saqua_ncbi = pd.read_table(\"C:/Users/eulle/Documents/besthits_ncbi_saqua.m8\")\n",
    "\n",
    "columns_csv = ['qseqid' ,'sseqid' ,'pident' ,'length','mismatch','gapopen','qstart' ,'qend' ,'sstart' ,'send' ,'evalue' ,'bitscore']\n",
    "row_values = pd.Series(df_saqua_ncbi.columns)\n",
    "df_saqua_ncbi = pd.DataFrame(np.vstack([df_saqua_ncbi.columns,df_saqua_ncbi]))\n",
    "df_saqua_ncbi.columns = columns_csv\n",
    "\n",
    "list_genkbank = list(df_saqua_ncbi[\"sseqid\"])\n",
    "list_genkbank = list(set(list_genkbank))\n",
    "p_regex = re.compile(r'^WP')\n",
    "refseq_id = [ii for ii in list_genkbank if p_regex.match(ii)]\n",
    "embl_id = [ii for ii in list_genkbank if not p_regex.match(ii)]   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Aqui são todas as bibliotecas e func utilizadas\n",
    "# Ter cuidado com o mapping_ids para quando for rodar n esquecer de mudar os parâmetros\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re,glob,os\n",
    "import urllib.parse\n",
    "import urllib.request\n",
    "from multiprocessing.pool import ThreadPool\n",
    "import urllib\n",
    "import pandas as pd\n",
    "import operator\n",
    "import functools\n",
    "from operator import methodcaller\n",
    "\n",
    "URL = 'https://www.uniprot.org/uploadlists/'\n",
    "\n",
    "def create_chunks(list_name, n):\n",
    "    for i in range(0, len(list_name), n):\n",
    "        yield list_name[i:i + n]\n",
    "        \n",
    "def mapping_ids(ids2map, source_fmt='ACC',target_fmt='KO_ID', output_fmt='tab'):\n",
    "    if hasattr(ids2map, 'pop'):\n",
    "            ids2map = ' '.join(ids2map)\n",
    "    params = {\n",
    "    'from': source_fmt,\n",
    "    'to': target_fmt,\n",
    "    'format': output_fmt,\n",
    "    'query': ids2map\n",
    "    }\n",
    "\n",
    "    data = urllib.parse.urlencode(params)\n",
    "    data = data.encode('utf-8')\n",
    "    req = urllib.request.Request(URL, data)\n",
    "    with urllib.request.urlopen(req) as f:\n",
    "        response = f.read().decode('utf-8')\n",
    "    return response\n",
    "\n",
    "def concat_dfs(path,extension):\n",
    "    extension = extension\n",
    "    all_filenames = [i for i in glob.glob('*.{}'.format(extension))]\n",
    "    \n",
    "    combined_dfs = pd.concat([pd.read_table(f) for f in all_filenames ])\n",
    "    \n",
    "    return combined_dfs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Aqui é feito o mapeamento de refseq e embl para uniprot\n",
    "\n",
    "import urllib.parse\n",
    "import urllib.request\n",
    "from multiprocessing.pool import ThreadPool\n",
    "import glob,os\n",
    "\n",
    "\n",
    "URL = 'https://www.uniprot.org/uploadlists/'\n",
    "\n",
    "def concat_dfs(path,extension):\n",
    "    extension = extension\n",
    "    all_filenames = [i for i in glob.glob('*.{}'.format(extension))]\n",
    "    \n",
    "    combined_dfs = pd.concat([pd.read_table(f) for f in all_filenames ])\n",
    "    \n",
    "    return combined_dfs\n",
    "\n",
    "\n",
    "def create_chunks(list_name, n):\n",
    "    for i in range(0, len(list_name), n):\n",
    "        yield list_name[i:i + n]\n",
    "\n",
    "def mapping_ids(ids2map, source_fmt='P_REFSEQ_AC',target_fmt='ACC', output_fmt='tab'):\n",
    "    if hasattr(ids2map, 'pop'):\n",
    "            ids2map = ' '.join(ids2map)\n",
    "    params = {\n",
    "    'from': source_fmt,\n",
    "    'to': target_fmt,\n",
    "    'format': output_fmt,\n",
    "    'query': ids2map\n",
    "    }\n",
    "\n",
    "    data = urllib.parse.urlencode(params)\n",
    "    data = data.encode('utf-8')\n",
    "    req = urllib.request.Request(URL, data)\n",
    "    with urllib.request.urlopen(req) as f:\n",
    "        response = f.read().decode('utf-8')\n",
    "    return response\n",
    "\n",
    "list_urls = []\n",
    "\n",
    "chunk_embl = list(create_chunks(embl_id,1000))\n",
    "chunk_refseq = list(create_chunks(refseq_id,1000))\n",
    "\n",
    "aux = ThreadPool(10).imap(mapping_ids,chunk_refseq)\n",
    "\n",
    "for index,ii in enumerate(aux): \n",
    "    with open('genkbank_ids/'+ str(index)+'_refseq'+'.csv', 'w') as file:\n",
    "        file.write(ii)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mapeando os embl_ids para ko\n",
    "\n",
    "concat_embl = concat_dfs(\"genkbank_ids/embl/\",\"csv\")\n",
    "# 274 rows duplicated -- EMBL\n",
    "concat_embl = concat_embl.reset_index(drop=True)\n",
    "concat_embl.columns = [\"EMBL\",\"Uniprot_ID\"]\n",
    "chunk_uniprotid = list(create_chunks(list(concat_embl[\"Uniprot_ID\"]),1000))\n",
    "\n",
    "pool_uniprot_ko_embl = ThreadPool(10).imap(mapping_ids,chunk_uniprotid)\n",
    "\n",
    "for index,ii in enumerate(pool_uniprot_ko_embl): \n",
    "    with open('uniprot_ko/'+ str(index)+'_ko'+'.csv', 'w') as file:\n",
    "        file.write(ii)\n",
    "#concat_embl.loc[concat_embl[\"EMBL\"].duplicated()]\n",
    "\n",
    "concat_embl_ko = concat_dfs(\"genkbank_ids/embl/uniprot_ko/\",\"csv\")\n",
    "concat_embl_ko = concat_embl_ko.reset_index(drop=True)\n",
    "concat_embl_ko.columns = [\"Uniprot_ID\",\"KO\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[WinError 2] O sistema não pode encontrar o arquivo especificado: '../../refseq/'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-67-1d759b10fb87>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     13\u001b[0m \u001b[1;31m#concat_embl.loc[concat_embl[\"EMBL\"].duplicated()]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     14\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 15\u001b[1;33m \u001b[0mconcat_refseq_ko\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mconcat_dfs\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mos\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mchdir\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"../../refseq/\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;34m\"csv\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     16\u001b[0m \u001b[0mconcat_refseq_ko\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mconcat_refseq_ko\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mreset_index\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdrop\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mTrue\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     17\u001b[0m \u001b[0mconcat_refseq_ko\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcolumns\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m[\u001b[0m\u001b[1;34m\"Uniprot_ID\"\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;34m\"KO\"\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mFileNotFoundError\u001b[0m: [WinError 2] O sistema não pode encontrar o arquivo especificado: '../../refseq/'"
     ]
    }
   ],
   "source": [
    "# Mapeando os refseq_ids para ko\n",
    "# lembrar de mudar esse concat\n",
    "\n",
    "concat_refseq = concat_dfs(os.chdir(\"./\"),\"csv\")\n",
    "\n",
    "concat_refseq = concat_refseq.reset_index(drop=True)\n",
    "concat_refseq.columns = [\"REFSEQ\",\"Uniprot_ID\"]\n",
    "chunk_uniprotid_refseq = list(create_chunks(list(concat_refseq[\"Uniprot_ID\"]),1000))\n",
    "\n",
    "pool_uniprot_ko_refseq = ThreadPool(10).imap(mapping_ids,chunk_uniprotid_refseq)\n",
    "\n",
    "for index,ii in enumerate(pool_uniprot_ko_refseq): \n",
    "    with open('uniprot_ko_refseq/'+ str(index)+'_ko_refseq'+'.csv', 'w') as file:\n",
    "        file.write(ii)\n",
    "# Da próxima vez que vc entrar só precisa ler essas linhas de baixo\n",
    "concat_refseq_ko = concat_dfs(os.chdir(\"uniprot_ko_refseq/\"),\"csv\")\n",
    "concat_refseq_ko = concat_refseq_ko.reset_index(drop=True)\n",
    "concat_refseq_ko.columns = [\"Uniprot_ID\",\"KO\"]\n",
    "\n",
    "concat_refseq_ko.to_csv(\"../uniprot_ko_refseq/df_ko_refseq.csv\",sep=\"\\t\",encoding=\"utf-8\",index=False)\n",
    "concat_embl_ko.to_csv(\"../../embl/uniprot_ko/df_ko_embl.csv\",sep=\"\\t\",encoding=\"utf-8\",index=False)\n",
    "\n",
    "## meter esses dois dfs no rstudio para mapear as vias\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Concatenação dos dados\n",
    "\n",
    "import pandas as pd\n",
    "df_genkbank = pd.read_csv(\"pathways_ncbixsaqua\")\n",
    "\n",
    "concat_embl = concat_dfs(os.chdir(\"./\"),\"csv\")\n",
    "# 274 rows duplicated -- EMBL\n",
    "concat_embl = concat_embl.reset_index(drop=True)\n",
    "concat_embl.columns = [\"genkbank\",\"Uniprot_ID\"]\n",
    "\n",
    "concat_refseq = concat_dfs(os.chdir(\"../refseq/\"),\"csv\")\n",
    "\n",
    "concat_refseq = concat_refseq.reset_index(drop=True)\n",
    "concat_refseq.columns = [\"genkbank\",\"Uniprot_ID\"]\n",
    "\n",
    "bigdata = pd.concat([concat_refseq,concat_embl])\n",
    "df_geral = pd.merge(bigdata,df_genkbank,on=\"Uniprot_ID\",how=\"left\")\n",
    "\n",
    "teta = df_geral[df_geral[\"Pathway\"].notna()]\n",
    "teta = teta.reset_index()\n",
    "teta = teta.drop_duplicated(\"genkbank\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'Pathway'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32m~\\Anaconda3\\lib\\site-packages\\pandas\\core\\indexes\\base.py\u001b[0m in \u001b[0;36mget_loc\u001b[1;34m(self, key, method, tolerance)\u001b[0m\n\u001b[0;32m   2896\u001b[0m             \u001b[1;32mtry\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 2897\u001b[1;33m                 \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   2898\u001b[0m             \u001b[1;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mpandas\\_libs\\index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;32mpandas\\_libs\\index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;32mpandas\\_libs\\hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;32mpandas\\_libs\\hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;31mKeyError\u001b[0m: 'Pathway'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-47-b8a81b38ef32>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[0mdf_saqua_ncbi\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmerge\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdf_saqua_ncbi\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mteta\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mleft_on\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m\"sseqid\"\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mright_on\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m\"genkbank\"\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mhow\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m\"left\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 2\u001b[1;33m \u001b[0mdf_saqua_ncbi\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mdf_saqua_ncbi\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mdf_saqua_ncbi\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m\"Pathway\"\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mnotna\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;31m# 25497\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      3\u001b[0m \u001b[0mdf_saqua_ncbi\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mdf_saqua_ncbi\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mreset_index\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[0mdf_saqua_ncbi\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mdrop\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m\"leveo_0\"\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;34m\"index\"\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0maxis\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      5\u001b[0m \u001b[1;31m#df_saqua_ncbi[df_saqua_ncbi[\"Pathway\"].notna()]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\lib\\site-packages\\pandas\\core\\frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m   2978\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mnlevels\u001b[0m \u001b[1;33m>\u001b[0m \u001b[1;36m1\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2979\u001b[0m                 \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_getitem_multilevel\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 2980\u001b[1;33m             \u001b[0mindexer\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   2981\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0mis_integer\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2982\u001b[0m                 \u001b[0mindexer\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m[\u001b[0m\u001b[0mindexer\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\lib\\site-packages\\pandas\\core\\indexes\\base.py\u001b[0m in \u001b[0;36mget_loc\u001b[1;34m(self, key, method, tolerance)\u001b[0m\n\u001b[0;32m   2897\u001b[0m                 \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2898\u001b[0m             \u001b[1;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 2899\u001b[1;33m                 \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_maybe_cast_indexer\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   2900\u001b[0m         \u001b[0mindexer\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_indexer\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mmethod\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mmethod\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mtolerance\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mtolerance\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2901\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mindexer\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mndim\u001b[0m \u001b[1;33m>\u001b[0m \u001b[1;36m1\u001b[0m \u001b[1;32mor\u001b[0m \u001b[0mindexer\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msize\u001b[0m \u001b[1;33m>\u001b[0m \u001b[1;36m1\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mpandas\\_libs\\index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;32mpandas\\_libs\\index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;32mpandas\\_libs\\hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;32mpandas\\_libs\\hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;31mKeyError\u001b[0m: 'Pathway'"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "df_saqua_ncbi = pd.merge(df_saqua_ncbi,teta,left_on=\"sseqid\",right_on=\"genkbank\",how=\"left\")\n",
    "df_saqua_ncbi.to_csv(\"../../saqua_ncbi_final.csv\",sep=\"\\t\",encoding=\"utf-8\",index=False)\n",
    "\n",
    "df_saqua_ncbi = df_saqua_ncbi[df_saqua_ncbi[\"Pathway\"].notna()] # 25497\n",
    "df_saqua_ncbi = df_saqua_ncbi.reset_index()\n",
    "df_saqua_ncbi = df_saqua_ncbi.drop(columns=[\"level_0\",\"index\"])\n",
    "\n",
    "#df_saqua_ncbi[df_saqua_ncbi[\"Pathway\"].notna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GABAergic synapse',\n",
       " 'Propanoate metabolism',\n",
       " 'Longevity regulating pathway',\n",
       " 'Central carbon metabolism in cancer',\n",
       " 'Huntington disease',\n",
       " 'Insulin signaling pathway',\n",
       " 'Ether lipid metabolism',\n",
       " 'Plant-pathogen interaction',\n",
       " 'Base excision repair',\n",
       " 'Riboflavin metabolism',\n",
       " 'Adipocytokine signaling pathway',\n",
       " 'Biosynthesis of secondary metabolites',\n",
       " 'HIF-1 signaling pathway',\n",
       " 'One carbon pool by folate',\n",
       " 'Nucleotide excision repair',\n",
       " 'Glucagon signaling pathway',\n",
       " 'Citrate cycle (TCA cycle)',\n",
       " 'Flagellar assembly',\n",
       " 'Viral carcinogenesis',\n",
       " 'Photosynthesis',\n",
       " 'RNA degradation',\n",
       " 'Proteoglycans in cancer',\n",
       " 'Necroptosis',\n",
       " 'Carbon metabolism',\n",
       " 'Cushing syndrome',\n",
       " 'Lipoic acid metabolism',\n",
       " 'Renal cell carcinoma',\n",
       " 'Arginine and proline metabolism',\n",
       " 'Parkinson disease',\n",
       " 'Axon regeneration',\n",
       " 'Aminobenzoate degradation',\n",
       " 'Type II diabetes mellitus',\n",
       " 'Pathways in cancer',\n",
       " 'Ubiquinone and other terpenoid-quinone biosynthesis',\n",
       " 'Biotin metabolism',\n",
       " 'Tryptophan metabolism',\n",
       " 'Biosynthesis of various secondary metabolites - part 2',\n",
       " 'Purine metabolism',\n",
       " 'Folate biosynthesis',\n",
       " 'Amino sugar and nucleotide sugar metabolism',\n",
       " 'Naphthalene degradation',\n",
       " 'Ascorbate and aldarate metabolism',\n",
       " 'Caprolactam degradation',\n",
       " 'Metabolic pathways',\n",
       " 'Fatty acid degradation',\n",
       " 'Histidine metabolism',\n",
       " 'Streptomycin biosynthesis',\n",
       " 'Arabinogalactan biosynthesis - Mycobacterium',\n",
       " '2-Oxocarboxylic acid metabolism',\n",
       " 'Glutamatergic synapse',\n",
       " 'Phospholipase D signaling pathway',\n",
       " 'FoxO signaling pathway',\n",
       " 'Peptidoglycan biosynthesis',\n",
       " 'Chlorocyclohexane and chlorobenzene degradation',\n",
       " 'Lysine degradation',\n",
       " 'Phosphonate and phosphinate metabolism',\n",
       " 'Mineral absorption',\n",
       " 'Tropane, piperidine and pyridine alkaloid biosynthesis',\n",
       " 'Biosynthesis of amino acids',\n",
       " 'Steroid hormone biosynthesis',\n",
       " 'Biofilm formation - Pseudomonas aeruginosa',\n",
       " 'Epithelial cell signaling in Helicobacter pylori infection',\n",
       " 'Prostate cancer',\n",
       " 'Glycine, serine and threonine metabolism',\n",
       " 'Phosphotransferase system (PTS)',\n",
       " 'Amoebiasis',\n",
       " 'Styrene degradation',\n",
       " 'Degradation of aromatic compounds',\n",
       " 'Nonribosomal peptide structures',\n",
       " 'Antigen processing and presentation',\n",
       " 'Ribosome',\n",
       " 'Monobactam biosynthesis',\n",
       " 'Pathogenic Escherichia coli infection',\n",
       " 'Type I diabetes mellitus',\n",
       " 'RNA transport',\n",
       " 'Terpenoid backbone biosynthesis',\n",
       " 'Toluene degradation',\n",
       " 'Tuberculosis',\n",
       " 'Cationic antimicrobial peptide (CAMP) resistance',\n",
       " 'Bacterial chemotaxis',\n",
       " 'AMPK signaling pathway',\n",
       " 'Geraniol degradation',\n",
       " 'Non-alcoholic fatty liver disease (NAFLD)',\n",
       " 'Cysteine and methionine metabolism',\n",
       " 'Metabolism of xenobiotics by cytochrome P450',\n",
       " 'Taurine and hypotaurine metabolism',\n",
       " 'Fat digestion and absorption',\n",
       " 'Legionellosis',\n",
       " 'Fluid shear stress and atherosclerosis',\n",
       " 'PI3K-Akt signaling pathway',\n",
       " 'Drug metabolism - cytochrome P450',\n",
       " 'Ethylbenzene degradation',\n",
       " 'Thermogenesis',\n",
       " 'Linoleic acid metabolism',\n",
       " 'Ferroptosis',\n",
       " 'Thiamine metabolism',\n",
       " 'D-Alanine metabolism',\n",
       " 'Lysine biosynthesis',\n",
       " 'Starch and sucrose metabolism',\n",
       " 'Glutathione metabolism',\n",
       " 'Alanine, aspartate and glutamate metabolism',\n",
       " 'Salmonella infection',\n",
       " 'Human papillomavirus infection',\n",
       " 'C5-Branched dibasic acid metabolism',\n",
       " 'Sulfur relay system',\n",
       " 'Isoquinoline alkaloid biosynthesis',\n",
       " 'Pyruvate metabolism',\n",
       " 'Homologous recombination',\n",
       " 'alpha-Linolenic acid metabolism',\n",
       " 'Selenocompound metabolism',\n",
       " 'Bacterial secretion system',\n",
       " 'Sulfur metabolism',\n",
       " 'Biofilm formation - Vibrio cholerae',\n",
       " 'MicroRNAs in cancer',\n",
       " 'D-Glutamine and D-glutamate metabolism',\n",
       " 'Glycerophospholipid metabolism',\n",
       " 'Meiosis - yeast',\n",
       " 'Estrogen signaling pathway',\n",
       " 'Apoptosis - fly',\n",
       " 'Fluorobenzoate degradation',\n",
       " 'Progesterone-mediated oocyte maturation',\n",
       " 'Chloroalkane and chloroalkene degradation',\n",
       " 'MAPK signaling pathway - fly',\n",
       " 'Biosynthesis of vancomycin group antibiotics',\n",
       " 'Protein export',\n",
       " 'Oxidative phosphorylation',\n",
       " 'Hepatocellular carcinoma',\n",
       " 'Pentose phosphate pathway',\n",
       " 'Pentose and glucuronate interconversions',\n",
       " 'Primary immunodeficiency',\n",
       " 'Galactose metabolism',\n",
       " 'MAPK signaling pathway - plant',\n",
       " 'RNA polymerase',\n",
       " 'Proximal tubule bicarbonate reclamation',\n",
       " 'Insulin resistance',\n",
       " 'Phenazine biosynthesis',\n",
       " 'Penicillin and cephalosporin biosynthesis',\n",
       " 'beta-Lactam resistance',\n",
       " 'DNA replication',\n",
       " 'Vancomycin resistance',\n",
       " 'Mismatch repair',\n",
       " 'Synthesis and degradation of ketone bodies',\n",
       " 'Choline metabolism in cancer',\n",
       " 'Fatty acid biosynthesis',\n",
       " 'Aminoacyl-tRNA biosynthesis',\n",
       " 'Ribosome biogenesis in eukaryotes',\n",
       " 'Polycyclic aromatic hydrocarbon degradation',\n",
       " 'Drug metabolism - other enzymes',\n",
       " 'Biofilm formation - Escherichia coli',\n",
       " 'Platinum drug resistance',\n",
       " 'Shigellosis',\n",
       " 'MAPK signaling pathway - yeast',\n",
       " 'Porphyrin and chlorophyll metabolism',\n",
       " 'Glucosinolate biosynthesis',\n",
       " 'Chagas disease (American trypanosomiasis)',\n",
       " 'Alzheimer disease',\n",
       " 'Butanoate metabolism',\n",
       " 'Phosphatidylinositol signaling system',\n",
       " 'Nitrogen metabolism',\n",
       " 'Amyotrophic lateral sclerosis (ALS)',\n",
       " 'IL-17 signaling pathway',\n",
       " 'Protein processing in endoplasmic reticulum',\n",
       " 'Glyoxylate and dicarboxylate metabolism',\n",
       " 'Biosynthesis of unsaturated fatty acids',\n",
       " 'Inositol phosphate metabolism',\n",
       " 'Glycolysis / Gluconeogenesis',\n",
       " 'Chemical carcinogenesis',\n",
       " 'NOD-like receptor signaling pathway',\n",
       " 'Thyroid hormone synthesis',\n",
       " 'Prodigiosin biosynthesis',\n",
       " 'Fructose and mannose metabolism',\n",
       " 'Peroxisome',\n",
       " 'African trypanosomiasis',\n",
       " 'Limonene and pinene degradation',\n",
       " 'Retinol metabolism',\n",
       " 'Lipopolysaccharide biosynthesis',\n",
       " 'Sphingolipid metabolism',\n",
       " 'Glycosphingolipid biosynthesis - globo and isoglobo series',\n",
       " 'Valine, leucine and isoleucine degradation',\n",
       " 'beta-Alanine metabolism',\n",
       " 'Nicotinate and nicotinamide metabolism',\n",
       " 'PPAR signaling pathway',\n",
       " 'Dioxin degradation',\n",
       " 'Polyketide sugar unit biosynthesis',\n",
       " 'Antifolate resistance',\n",
       " 'Longevity regulating pathway - multiple species',\n",
       " 'Biosynthesis of ansamycins',\n",
       " 'Vitamin B6 metabolism',\n",
       " 'Cardiac muscle contraction',\n",
       " 'Arachidonic acid metabolism',\n",
       " 'Phenylalanine metabolism',\n",
       " 'Insect hormone biosynthesis',\n",
       " 'Cyanoamino acid metabolism',\n",
       " 'Pyrimidine metabolism',\n",
       " 'Acarbose and validamycin biosynthesis',\n",
       " 'Glycerolipid metabolism',\n",
       " 'Carbon fixation pathways in prokaryotes',\n",
       " 'Pantothenate and CoA biosynthesis',\n",
       " 'Novobiocin biosynthesis',\n",
       " 'Carbapenem biosynthesis',\n",
       " 'Biosynthesis of siderophore group nonribosomal peptides',\n",
       " 'Longevity regulating pathway - worm',\n",
       " 'Arginine biosynthesis',\n",
       " 'Fatty acid metabolism',\n",
       " 'Microbial metabolism in diverse environments',\n",
       " 'Benzoate degradation',\n",
       " 'Methane metabolism',\n",
       " 'Quorum sensing',\n",
       " 'Valine, leucine and isoleucine biosynthesis',\n",
       " 'Th17 cell differentiation',\n",
       " 'Tyrosine metabolism',\n",
       " 'Other glycan degradation',\n",
       " 'Cell cycle - Caulobacter',\n",
       " 'Xylene degradation',\n",
       " 'Two-component system',\n",
       " 'Pertussis',\n",
       " 'Carbon fixation in photosynthetic organisms',\n",
       " 'Thyroid hormone signaling pathway',\n",
       " 'ABC transporters',\n",
       " 'Atrazine degradation',\n",
       " 'Phenylalanine, tyrosine and tryptophan biosynthesis']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_saqua_ncbi = pd.read_table(\"saqua_ncbi_final.csv\")\n",
    "df_saqua_ncbi = df_saqua_ncbi[df_saqua_ncbi[\"Pathway\"].notna()]\n",
    "pathways = df_saqua_ncbi[\"Pathway\"]\n",
    "pathways = list(pathways)\n",
    "pathways = map(methodcaller(\"split\",\";\"),pathways)\n",
    "pathways = functools.reduce(operator.iconcat, pathways, [])\n",
    "pathways = list(set(pathways))\n",
    "pathways\n",
    "# 221 vias distintas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\eulle\\Anaconda3\\lib\\site-packages\\pandas\\core\\strings.py:1843: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.\n",
      "  return func(self, *args, **kwargs)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'Necroptosis': 90,\n",
       " 'Sphingolipid metabolism': 7,\n",
       " 'Ether lipid metabolism': 117,\n",
       " 'Quorum sensing': 897,\n",
       " 'Glycerolipid metabolism': 164,\n",
       " 'Biosynthesis of vancomycin group antibiotics': 12,\n",
       " 'Bacterial secretion system': 337,\n",
       " 'Tuberculosis': 29,\n",
       " 'Peroxisome': 138,\n",
       " 'Renal cell carcinoma': 1,\n",
       " 'MAPK signaling pathway - yeast': 6,\n",
       " 'Cushing syndrome': 1,\n",
       " 'Non-alcoholic fatty liver disease (NAFLD)': 0,\n",
       " 'DNA replication': 148,\n",
       " 'Salmonella infection': 92,\n",
       " 'Viral carcinogenesis': 2,\n",
       " 'Biotin metabolism': 470,\n",
       " 'PPAR signaling pathway': 154,\n",
       " 'Biosynthesis of secondary metabolites': 7023,\n",
       " 'Axon regeneration': 1,\n",
       " 'Phosphonate and phosphinate metabolism': 138,\n",
       " 'Carbapenem biosynthesis': 364,\n",
       " 'Central carbon metabolism in cancer': 512,\n",
       " 'Dioxin degradation': 10,\n",
       " 'Lysine biosynthesis': 417,\n",
       " 'Plant-pathogen interaction': 112,\n",
       " 'Thyroid hormone signaling pathway': 117,\n",
       " 'AMPK signaling pathway': 31,\n",
       " 'Biofilm formation - Pseudomonas aeruginosa': 409,\n",
       " 'MAPK signaling pathway - fly': 104,\n",
       " 'Citrate cycle (TCA cycle)': 0,\n",
       " 'Tyrosine metabolism': 175,\n",
       " 'Longevity regulating pathway': 434,\n",
       " 'One carbon pool by folate': 259,\n",
       " 'Nucleotide excision repair': 24,\n",
       " 'Phenazine biosynthesis': 3,\n",
       " 'Chemical carcinogenesis': 28,\n",
       " 'Proteoglycans in cancer': 5,\n",
       " 'Valine, leucine and isoleucine degradation': 524,\n",
       " 'Two-component system': 2488,\n",
       " 'Microbial metabolism in diverse environments': 4494,\n",
       " 'Cell cycle - Caulobacter': 738,\n",
       " 'Pyruvate metabolism': 814,\n",
       " 'Glucosinolate biosynthesis': 360,\n",
       " 'Selenocompound metabolism': 54,\n",
       " 'Glyoxylate and dicarboxylate metabolism': 867,\n",
       " 'Tryptophan metabolism': 164,\n",
       " 'Methane metabolism': 252,\n",
       " 'Glycerophospholipid metabolism': 156,\n",
       " 'Lysine degradation': 127,\n",
       " 'Longevity regulating pathway - worm': 317,\n",
       " 'Histidine metabolism': 210,\n",
       " 'RNA degradation': 121,\n",
       " 'Lipopolysaccharide biosynthesis': 200,\n",
       " 'Arginine and proline metabolism': 539,\n",
       " 'Chloroalkane and chloroalkene degradation': 34,\n",
       " 'Insulin signaling pathway': 13,\n",
       " 'Arachidonic acid metabolism': 27,\n",
       " 'Streptomycin biosynthesis': 123,\n",
       " 'Benzoate degradation': 772,\n",
       " '2-Oxocarboxylic acid metabolism': 1706,\n",
       " 'Lipoic acid metabolism': 5,\n",
       " 'Thiamine metabolism': 90,\n",
       " 'Phospholipase D signaling pathway': 1,\n",
       " 'Legionellosis': 33,\n",
       " 'Ascorbate and aldarate metabolism': 124,\n",
       " 'Mineral absorption': 12,\n",
       " 'Glutathione metabolism': 137,\n",
       " 'Primary immunodeficiency': 2,\n",
       " 'Xylene degradation': 28,\n",
       " 'Antigen processing and presentation': 5,\n",
       " 'Amino sugar and nucleotide sugar metabolism': 579,\n",
       " 'Toluene degradation': 222,\n",
       " 'Linoleic acid metabolism': 1,\n",
       " 'PI3K-Akt signaling pathway': 8,\n",
       " 'GABAergic synapse': 575,\n",
       " 'NOD-like receptor signaling pathway': 86,\n",
       " 'beta-Alanine metabolism': 42,\n",
       " 'Nitrogen metabolism': 252,\n",
       " 'Parkinson disease': 84,\n",
       " 'Oxidative phosphorylation': 1369,\n",
       " 'Propanoate metabolism': 557,\n",
       " 'Fatty acid biosynthesis': 715,\n",
       " 'Pathogenic Escherichia coli infection': 7,\n",
       " 'Glucagon signaling pathway': 21,\n",
       " 'Pentose phosphate pathway': 369,\n",
       " 'Limonene and pinene degradation': 22,\n",
       " 'Longevity regulating pathway - multiple species': 227,\n",
       " 'Th17 cell differentiation': 5,\n",
       " 'RNA transport': 8,\n",
       " 'Epithelial cell signaling in Helicobacter pylori infection': 9,\n",
       " 'Arginine biosynthesis': 1279,\n",
       " 'Nicotinate and nicotinamide metabolism': 481,\n",
       " 'Estrogen signaling pathway': 5,\n",
       " 'Type I diabetes mellitus': 21,\n",
       " 'Purine metabolism': 1467,\n",
       " 'Pyrimidine metabolism': 441,\n",
       " 'Drug metabolism - other enzymes': 321,\n",
       " 'Bacterial chemotaxis': 52,\n",
       " 'Riboflavin metabolism': 306,\n",
       " 'Sulfur metabolism': 747,\n",
       " 'alpha-Linolenic acid metabolism': 5,\n",
       " 'Drug metabolism - cytochrome P450': 37,\n",
       " 'Phenylalanine metabolism': 207,\n",
       " 'Cardiac muscle contraction': 5,\n",
       " 'Retinol metabolism': 17,\n",
       " 'Polycyclic aromatic hydrocarbon degradation': 97,\n",
       " 'Novobiocin biosynthesis': 10,\n",
       " 'Metabolism of xenobiotics by cytochrome P450': 37,\n",
       " 'Chagas disease (American trypanosomiasis)': 0,\n",
       " 'Chlorocyclohexane and chlorobenzene degradation': 232,\n",
       " 'Adipocytokine signaling pathway': 4,\n",
       " 'Amoebiasis': 2,\n",
       " 'Ubiquinone and other terpenoid-quinone biosynthesis': 23,\n",
       " 'Cyanoamino acid metabolism': 25,\n",
       " 'Ferroptosis': 2,\n",
       " 'C5-Branched dibasic acid metabolism': 624,\n",
       " 'Isoquinoline alkaloid biosynthesis': 14,\n",
       " 'Metabolic pathways': 15993,\n",
       " 'MicroRNAs in cancer': 503,\n",
       " 'Butanoate metabolism': 817,\n",
       " 'African trypanosomiasis': 2,\n",
       " 'Type II diabetes mellitus': 2,\n",
       " 'Glycosphingolipid biosynthesis - globo and isoglobo series': 3,\n",
       " 'Carbon fixation pathways in prokaryotes': 768,\n",
       " 'Thermogenesis': 15,\n",
       " 'Porphyrin and chlorophyll metabolism': 354,\n",
       " 'Cationic antimicrobial peptide (CAMP) resistance': 0,\n",
       " 'Steroid hormone biosynthesis': 4,\n",
       " 'HIF-1 signaling pathway': 30,\n",
       " 'Degradation of aromatic compounds': 618,\n",
       " 'Shigellosis': 8,\n",
       " 'Choline metabolism in cancer': 1,\n",
       " 'Phosphotransferase system (PTS)': 0,\n",
       " 'Mismatch repair': 147,\n",
       " 'Vancomycin resistance': 516,\n",
       " 'Alanine, aspartate and glutamate metabolism': 890,\n",
       " 'Nonribosomal peptide structures': 2,\n",
       " 'Protein export': 379,\n",
       " 'Sulfur relay system': 192,\n",
       " 'Other glycan degradation': 2,\n",
       " 'ABC transporters': 1813,\n",
       " 'Homologous recombination': 202,\n",
       " 'Platinum drug resistance': 124,\n",
       " 'Meiosis - yeast': 112,\n",
       " 'Ribosome': 2101,\n",
       " 'Fluorobenzoate degradation': 35,\n",
       " 'IL-17 signaling pathway': 5,\n",
       " 'Biofilm formation - Escherichia coli': 263,\n",
       " 'Hepatocellular carcinoma': 20,\n",
       " 'Terpenoid backbone biosynthesis': 194,\n",
       " 'D-Glutamine and D-glutamate metabolism': 530,\n",
       " 'Human papillomavirus infection': 2,\n",
       " 'Acarbose and validamycin biosynthesis': 13,\n",
       " 'Alzheimer disease': 12,\n",
       " 'Pantothenate and CoA biosynthesis': 1617,\n",
       " 'Naphthalene degradation': 23,\n",
       " 'Biosynthesis of amino acids': 3529,\n",
       " 'Aminobenzoate degradation': 284,\n",
       " 'Synthesis and degradation of ketone bodies': 94,\n",
       " 'Biosynthesis of unsaturated fatty acids': 108,\n",
       " 'Biosynthesis of siderophore group nonribosomal peptides': 117,\n",
       " 'Pentose and glucuronate interconversions': 212,\n",
       " 'Starch and sucrose metabolism': 50,\n",
       " 'Biosynthesis of ansamycins': 220,\n",
       " 'Pertussis': 8,\n",
       " 'Biofilm formation - Vibrio cholerae': 459,\n",
       " 'Flagellar assembly': 80,\n",
       " 'Carbon metabolism': 1764,\n",
       " 'Prodigiosin biosynthesis': 353,\n",
       " 'Fat digestion and absorption': 79,\n",
       " 'Inositol phosphate metabolism': 219,\n",
       " 'Insulin resistance': 4,\n",
       " 'MAPK signaling pathway - plant': 129,\n",
       " 'Progesterone-mediated oocyte maturation': 5,\n",
       " 'Insect hormone biosynthesis': 3,\n",
       " 'FoxO signaling pathway': 113,\n",
       " 'Glycine, serine and threonine metabolism': 474,\n",
       " 'Tropane, piperidine and pyridine alkaloid biosynthesis': 14,\n",
       " 'Biosynthesis of various secondary metabolites - part 2': 90,\n",
       " 'D-Alanine metabolism': 106,\n",
       " 'Photosynthesis': 672,\n",
       " 'Monobactam biosynthesis': 318,\n",
       " 'Geraniol degradation': 32,\n",
       " 'Glycolysis / Gluconeogenesis': 197,\n",
       " 'Taurine and hypotaurine metabolism': 74,\n",
       " 'Aminoacyl-tRNA biosynthesis': 280,\n",
       " 'Galactose metabolism': 121,\n",
       " 'Proximal tubule bicarbonate reclamation': 505,\n",
       " 'Antifolate resistance': 233,\n",
       " 'Fatty acid degradation': 168,\n",
       " 'Protein processing in endoplasmic reticulum': 5,\n",
       " 'Cysteine and methionine metabolism': 769,\n",
       " 'RNA polymerase': 318,\n",
       " 'Thyroid hormone synthesis': 28,\n",
       " 'Fluid shear stress and atherosclerosis': 113,\n",
       " 'Carbon fixation in photosynthetic organisms': 315,\n",
       " 'Valine, leucine and isoleucine biosynthesis': 1223,\n",
       " 'Vitamin B6 metabolism': 134,\n",
       " 'Phenylalanine, tyrosine and tryptophan biosynthesis': 141,\n",
       " 'Arabinogalactan biosynthesis - Mycobacterium': 4,\n",
       " 'Phosphatidylinositol signaling system': 97,\n",
       " 'Amyotrophic lateral sclerosis (ALS)': 0,\n",
       " 'Polyketide sugar unit biosynthesis': 14,\n",
       " 'Apoptosis - fly': 3,\n",
       " 'Glutamatergic synapse': 575,\n",
       " 'Fatty acid metabolism': 871,\n",
       " 'Peptidoglycan biosynthesis': 574,\n",
       " 'Ribosome biogenesis in eukaryotes': 11,\n",
       " 'Caprolactam degradation': 236,\n",
       " 'Penicillin and cephalosporin biosynthesis': 1,\n",
       " 'Prostate cancer': 5,\n",
       " 'Huntington disease': 136,\n",
       " 'Folate biosynthesis': 697,\n",
       " 'Pathways in cancer': 26,\n",
       " 'Styrene degradation': 146,\n",
       " 'Atrazine degradation': 66,\n",
       " 'Base excision repair': 82,\n",
       " 'Ethylbenzene degradation': 5,\n",
       " 'Fructose and mannose metabolism': 193,\n",
       " 'beta-Lactam resistance': 326}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pathways_count = {}\n",
    "for key in pathways:\n",
    "    aux = df_saqua_ncbi[df_saqua_ncbi[\"Pathway\"].str.contains(key)]\n",
    "    pathways_count[key] = aux[\"Pathway\"].count()\n",
    "pathways_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'pathways_count' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-36-c2758cdd8022>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      8\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      9\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 10\u001b[1;33m \u001b[0mx\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mpathways_count\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m\"Pathway\"\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;34m\"Count\"\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mindex\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mlist\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mpathways_count\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mkeys\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     11\u001b[0m \u001b[0mx\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mx\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msort_values\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mby\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m\"Count\"\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mascending\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mFalse\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     12\u001b[0m \u001b[0max\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mx\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mplot\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mbarh\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfigsize\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;36m500\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;36m200\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mcolor\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'#86bf91'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mzorder\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m2\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mwidth\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m0.95\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'pathways_count' is not defined"
     ]
    }
   ],
   "source": [
    "# Gerando o plot das vias\n",
    "# Como o num de vias é muito grande, vou repartir em dois gráficos\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.ticker import StrMethodFormatter\n",
    "\n",
    "pathways_1 = [0:111]\n",
    "pathways_2 = [111:221]\n",
    "\n",
    "x = pd.DataFrame(pathways_count.items(),columns=[\"Pathway\",\"Count\"],index=list(pathways_count.keys()))\n",
    "x = x.sort_values(by=[\"Count\"],ascending=False)\n",
    "ax = x.plot.barh(figsize=(500, 200), color='#86bf91', zorder=2, width=0.95)\n",
    "ax.set_alpha(0.8)\n",
    "ax.set_title(\"Saqua x NCBI\", fontsize=15)\n",
    "ax.set_xlabel(\"Count of reads per pathway\", fontsize=15);\n",
    "ax.set_xticks([0,500,1300,2000,3000,4200])\n",
    "\n",
    "for i in ax.patches:  \n",
    "    ax.text(i.get_width()+.1, i.get_y()+.60, \\\n",
    "            str(round((i.get_width()), 2)), fontsize=10, color='dimgrey')\n",
    "\n",
    "# invert for largest on top \n",
    "ax.invert_yaxis()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'pd' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-1-a61c0870245f>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[1;31m# Verificar quais são as vias que foram mapeadas tanto no NCBI quanto no Biosurfdb\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      2\u001b[0m \u001b[1;31m# As vias do ncbi foram geradas em uma célula anterior\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 3\u001b[1;33m \u001b[0mdf_saqua_biosurfdb\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mread_table\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"saqua_biosurfdb_final.csv\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      4\u001b[0m \u001b[0mdf_saqua_biosurfdb\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mdf_saqua_biosurfdb\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mdf_saqua_biosurfdb\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m\"Pathway\"\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mnotna\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      5\u001b[0m \u001b[0mpathways_biosufdb\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mdf_saqua_biosurfdb\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m\"Pathway\"\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'pd' is not defined"
     ]
    }
   ],
   "source": [
    "# Verificar quais são as vias que foram mapeadas tanto no NCBI quanto no Biosurfdb\n",
    "# As vias do ncbi foram geradas em uma célula anterior\n",
    "df_saqua_biosurfdb = pd.read_table(\"saqua_biosurfdb_final.csv\")\n",
    "df_saqua_biosurfdb = df_saqua_biosurfdb[df_saqua_biosurfdb[\"Pathway\"].notna()]\n",
    "pathways_biosufdb = df_saqua_biosurfdb[\"Pathway\"]\n",
    "pathways_biosufdb = list(pathways_biosufdb)\n",
    "pathways_biosufdb = map(methodcaller(\"split\",\";\"),pathways_biosufdb)\n",
    "pathways_biosufdb = functools.reduce(operator.iconcat, pathways_biosufdb, [])\n",
    "pathways_biosufdb = list(set(pathways_biosufdb))\n",
    "\n",
    "bs_bd = list(set(pathways).intersection(pathways_biosufdb))\n",
    "\n",
    "set(pathways_biosufdb) - set(bs_bd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Carbapenem biosynthesis',\n",
       " 'Geraniol degradation',\n",
       " 'Fluid shear stress and atherosclerosis',\n",
       " 'Biosynthesis of amino acids',\n",
       " 'Pathogenic Escherichia coli infection',\n",
       " 'Valine, leucine and isoleucine biosynthesis',\n",
       " 'Pertussis',\n",
       " 'Drug metabolism - cytochrome P450',\n",
       " 'Dioxin degradation',\n",
       " 'Proximal tubule bicarbonate reclamation',\n",
       " 'Tryptophan metabolism',\n",
       " 'D-Alanine metabolism',\n",
       " 'Base excision repair',\n",
       " 'Primary immunodeficiency',\n",
       " 'Monobactam biosynthesis',\n",
       " 'Choline metabolism in cancer',\n",
       " 'Apoptosis - fly',\n",
       " 'Ribosome biogenesis in eukaryotes',\n",
       " 'Antifolate resistance',\n",
       " 'Insulin signaling pathway',\n",
       " 'Chloroalkane and chloroalkene degradation',\n",
       " 'Peptidoglycan biosynthesis',\n",
       " 'Necroptosis',\n",
       " 'Bacterial chemotaxis',\n",
       " 'Butanoate metabolism',\n",
       " 'Epithelial cell signaling in Helicobacter pylori infection',\n",
       " 'Glycosphingolipid biosynthesis - globo and isoglobo series',\n",
       " 'Chagas disease (American trypanosomiasis)',\n",
       " 'Synthesis and degradation of ketone bodies',\n",
       " 'Glutamatergic synapse',\n",
       " 'Polyketide sugar unit biosynthesis',\n",
       " 'Cationic antimicrobial peptide (CAMP) resistance',\n",
       " 'Quorum sensing',\n",
       " 'Styrene degradation',\n",
       " 'Viral carcinogenesis',\n",
       " 'Phosphonate and phosphinate metabolism',\n",
       " 'GABAergic synapse',\n",
       " 'Caprolactam degradation',\n",
       " 'Amino sugar and nucleotide sugar metabolism',\n",
       " 'Phenylalanine, tyrosine and tryptophan biosynthesis',\n",
       " 'Carbon metabolism',\n",
       " 'Vancomycin resistance',\n",
       " 'Glycerolipid metabolism',\n",
       " 'Ascorbate and aldarate metabolism',\n",
       " 'Non-alcoholic fatty liver disease (NAFLD)',\n",
       " 'Cushing syndrome',\n",
       " 'Fructose and mannose metabolism',\n",
       " 'PI3K-Akt signaling pathway',\n",
       " 'Biosynthesis of ansamycins',\n",
       " 'Glutathione metabolism',\n",
       " 'Protein export']"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pathways[0:51]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
